### DISCUSSION TITLE
Allow specifying options to Shiki's highlighter to create code diffs/focus mode

### DISCUSSION THREAD

**Discussion by innocenzi** [withastro/roadmap#346] [Proposal]
Currently, Shiki's [`codeToHtml`](https://github.com/withastro/astro/blob/a2e8e76c303e8d6f39c24c122905a10f06907997/packages/markdown/remark/src/remark-shiki.ts#L63) is called without an opportunity in userland to pass some custom `lineOptions`.

[On another project](https://github.com/hybridly/hybridly/blob/0.x/docs/.vitepress/shiki-tags/highlighter.ts#L8), I have some processing code based on Shiki, that receives the code block's content and returns `lineOptions` and an updated code block. 

This allows me to apply classes on lines depending on what they contain, so I can extend Shiki's functionality (eg. to add colored diff or focus mode):

![image](https://user-images.githubusercontent.com/16060559/196239656-a561dc98-e2f1-49e6-acca-f852e20b6e5f.png)

![CleanShot 2022-10-17 at 19 08 40](https://user-images.githubusercontent.com/16060559/196240119-d1230733-1c37-45e7-98d8-ff86549a882b.gif)

I would like to suggest finding an API that would allow such customization on userland. I wanted to PR directly, but I was unsure of what it could look like, since Astro supports `prism` as well. 

For reference, the project from which these screenshots come is a VitePress project, which accepts an [`highlighter`](https://github.com/hybridly/hybridly/blob/0.x/docs/.vitepress/shiki-tags/highlighter.ts#L8) function in its config.

Any suggestion to achieve this with Astro is welcome :)

**Comment by innocenzi** [comment]
I made https://www.npmjs.com/package/shiki-processor, which bundles the code of my other project. The question of integrating it still stands though.

**Comment by youssefm** [comment]
Would absolutely love to see this now that Shiki has first class support for line options

**Comment by nmattia** [comment]
I'd also like to have something like this! I don't see any easy workarounds for custom classes per-line without this.

Another option would be to add support for this based on markdown metadata, similar to what the astro [docs](https://github.com/withastro/docs/blob/afb374ae7796b94f58192512d6bfe218847fa500/integrations/astro-code-snippets.ts#L193-L200) repo does (without the need for an extra component).

The [`codeToHtml` call](https://github.com/withastro/astro/blob/8b7cb64dadfca93c65d62df54754633d398cb2ed/packages/markdown/remark/src/remark-shiki.ts#L63) could be augmented with `lineOptions` derived from the node's meta:

```diff
-let html = highlighter!.codeToHtml(node.value, { lang });
+let lineOptions = parseLineOptions(code.meta)
+let html = highlighter!.codeToHtml(node.value, { lang, lineOptions });
```

where `parseLineOptions` would do something like this:
```
parseLineOptions("ins={2,4}")
> [{ line: 2, classes: ["ins"]}, { line: 4, classes: ["ins"]}]
parseLineOptions("ins={2-4} highlight={3}")
> [{ line: 2, classes: ["ins"]}, { line: 3, classes: ["ins","highlight"]}, { line: 4, classes: ["ins"]}]
```

Same as @innocenzi:
>  I wanted to PR directly, but I was unsure of what it could look like, since Astro supports prism as well.

@JuanM04 @FredKSchott looks like you've contributed most to `remark-shiki`, thoughts?

**Comment by xavdid** [comment]
+1, would be great to be able to highlight specific lines in a markdown codeblock without extra fuss. Seems like Shiki added support in https://github.com/shikijs/shiki/pull/259. It might be possible with Prism, but I haven't messed with that yet. 

````
```js {1-3,5}
// .. some js code with lines 1,2,3, and 5 highlighted
```
````
